File: frontend/mobile/lib/models/debt.dart
class Debt {
  final int id;
  final int customerId;
  final String customerName;
  final String orderCode;
  final double totalAmount;
  final double paidAmount;
  final double remainingAmount;
  final String status;
  final String createdAt;
  final String? dueDate;
  final List<DebtPayment>? payments;

  Debt({
    required this.id,
    required this.customerId,
    required this.customerName,
    required this.orderCode,
    required this.totalAmount,
    required this.paidAmount,
    required this.remainingAmount,
    required this.status,
    required this.createdAt,
    this.dueDate,
    this.payments,
  });

  factory Debt.fromJson(Map<String, dynamic> json) {
    return Debt(
      id: _parseInt(json['id']),
      customerId: _parseInt(json['customer_id']),
      customerName: json['customer_name'] ?? '',
      orderCode: json['order_code'] ?? '',
      totalAmount: _parseDouble(json['total_amount']),
      paidAmount: _parseDouble(json['paid_amount']),
      remainingAmount: _parseDouble(json['remaining_amount']),
      status: json['status'] ?? 'PENDING',
      createdAt: json['created_at'] ?? '',
      dueDate: json['due_date'],
      payments: json['payments'] != null
          ? (json['payments'] as List)
              .map((i) => DebtPayment.fromJson(i))
              .toList()
          : null,
    );
  }

  static double _parseDouble(dynamic value) {
    if (value == null) return 0.0;
    if (value is num) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? 0.0;
    return 0.0;
  }

  static int _parseInt(dynamic value) {
    if (value == null) return 0;
    if (value is num) return value.toInt();
    if (value is String) return int.tryParse(value) ?? 0;
    return 0;
  }
}

class DebtPayment {
  final int id;
  final double paymentAmount;
  final String paymentMethod;
  final String? referenceNumber;
  final String? notes;
  final String createdAt;

  DebtPayment({
    required this.id,
    required this.paymentAmount,
    required this.paymentMethod,
    this.referenceNumber,
    this.notes,
    required this.createdAt,
  });

  factory DebtPayment.fromJson(Map<String, dynamic> json) {
    return DebtPayment(
      id: _parseInt(json['id']),
      paymentAmount: _parseDouble(json['payment_amount']),
      paymentMethod: json['payment_method'] ?? 'CASH',
      referenceNumber: json['reference_number'],
      notes: json['notes'],
      createdAt: json['created_at'] ?? '',
    );
  }

  static double _parseDouble(dynamic value) {
    if (value == null) return 0.0;
    if (value is num) return value.toDouble();
    if (value is String) return double.tryParse(value) ?? 0.0;
    return 0.0;
  }

  static int _parseInt(dynamic value) {
    if (value == null) return 0;
    if (value is num) return value.toInt();
    if (value is String) return int.tryParse(value) ?? 0;
    return 0;
  }
}

File: backend/domain/entities/debt.py
"""
Debt entity - Customer debts and payments
"""
from datetime import date, datetime
from typing import Optional
from dataclasses import dataclass
from decimal import Decimal
from enum import Enum


class DebtStatus(str, Enum):
    """Debt status enumeration"""
    PENDING = "PENDING"
    PARTIAL = "PARTIAL"
    PAID = "PAID"
    OVERDUE = "OVERDUE"


class DebtPaymentMethod(str, Enum):
    """Debt payment method enumeration"""
    CASH = "CASH"
    BANK_TRANSFER = "BANK_TRANSFER"
    OTHER = "OTHER"


@dataclass
class DebtPayment:
    """
    Debt payment entity.
    
    Business Rules:
    - payment_amount must be > 0
    - Updates parent Debt.paid_amount
    """
    id: Optional[int]
    debt_id: int
    payment_amount: Decimal
    created_by: int
    payment_method: DebtPaymentMethod = DebtPaymentMethod.CASH
    payment_date: datetime = None
    reference_number: Optional[str] = None
    notes: Optional[str] = None
    created_at: Optional[datetime] = None
    
    def __post_init__(self):
        """Validate payment data"""
        if self.payment_amount <= 0:
            raise ValueError("Payment amount must be greater than 0")
        
        if self.payment_date is None:
            self.payment_date = datetime.utcnow()


@dataclass
class Debt:
    """
    Debt entity representing customer debts.
    
    Business Rules:
    - Auto-created when ORDER.debt_amount > 0
    - remaining_amount = debt_amount - paid_amount
    - status auto-updated based on payments and due_date
    - Enforces CUSTOMER.credit_limit
    """
    id: Optional[int]
    owner_id: int
    customer_id: int
    order_id: int
    debt_amount: Decimal
    paid_amount: Decimal = Decimal("0")
    remaining_amount: Decimal = Decimal("0")
    due_date: Optional[date] = None
    status: DebtStatus = DebtStatus.PENDING
    notes: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    
    def __post_init__(self):
        """Validate debt data"""
        if self.debt_amount <= 0:
            raise ValueError("Debt amount must be greater than 0")
        
        if self.paid_amount < 0:
            raise ValueError("Paid amount cannot be negative")
        
        if self.paid_amount > self.debt_amount:
            raise ValueError("Paid amount cannot exceed debt amount")
        
        # Sync remaining_amount if not set
        if self.remaining_amount == Decimal("0") and self.debt_amount > 0:
            self.remaining_amount = self.debt_amount - self.paid_amount
    
    def get_remaining_amount(self) -> Decimal:
        """Get remaining debt amount"""
        return self.remaining_amount
    
    def recalculate_remaining_amount(self):
        """Recalculate and update remaining_amount field"""
        self.remaining_amount = self.debt_amount - self.paid_amount
    
    def is_fully_paid(self) -> bool:
        """Check if debt is fully paid"""
        return self.get_remaining_amount() <= 0
    
    def is_overdue(self) -> bool:
        """Check if debt is overdue"""
        if self.due_date is None:
            return False
        return date.today() > self.due_date and not self.is_fully_paid()
    
    def record_payment(self, payment: DebtPayment):
        """
        Record a payment for this debt.
        
        Args:
            payment: DebtPayment object
        
        Raises:
            ValueError: If payment exceeds remaining amount
        """
        remaining = self.get_remaining_amount()
        
        if payment.payment_amount > remaining:
            raise ValueError(f"Payment amount {payment.payment_amount} exceeds remaining debt {remaining}")
        
        self.paid_amount += payment.payment_amount
        self.recalculate_remaining_amount()
        self.update_status()
    
    def update_status(self):
        """Update debt status based on paid amount and due date"""
        if self.is_fully_paid():
            self.status = DebtStatus.PAID
        elif self.paid_amount > 0:
            self.status = DebtStatus.PARTIAL
        elif self.is_overdue():
            self.status = DebtStatus.OVERDUE
        else:
            self.status = DebtStatus.PENDING
    
    def days_overdue(self) -> int:
        """Get number of days overdue (0 if not overdue)"""
        if not self.is_overdue():
            return 0
        
        delta = date.today() - self.due_date
        return delta.days
